### PR TITLE
Integrate burrego inside the policy-evaluator

### PR DESCRIPTION
**Pull request by ereslibre**
Initial integration. Many bits and pieces should be polished.

Related: https://github.com/kubewarden/policy-evaluator/issues/15

### CODE REVIEW

File: crates/burrego/.github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,127 @@
+name: burrego release
```
Comment (flavio) [inline crates/burrego/.github/workflows/release.yml:1]: I think we can remove this GitHub action

File: crates/burrego/.github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,77 @@
+on: [push, pull_request]
```
Comment (flavio) [inline crates/burrego/.github/workflows/tests.yml:1]: I wonder if this GitHub action is ran, that would be nice

File: crates/burrego/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+anyhow = "1.0.40"
+clap = "2.33.3"
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0.64"
+gtmpl = { git = "https://github.com/flavio/gtmpl-rust.git", branch = "printf-handle-array-and-map" }
```
Comment (flavio) [inline crates/burrego/Cargo.toml:14]: I moved this fork under the `kubewarden` org

File: crates/burrego/src/opa/evaluation_result.rs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+pub struct EvaluationResult {
+    pub result: bool,
+}
```
Comment (flavio) [inline crates/burrego/src/opa/evaluation_result.rs:6]: I think this is limiting. Violations can return also explanation messages about why something has been rejected

File: crates/burrego/src/opa/wasm.rs [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+            move |mut caller: Caller<'_, Option<StackHelper>>, addr: i32| {
+                let stack_helper = caller.data().unwrap();
+                let msg = stack_helper
+                    .pull_json(caller.as_context_mut(), &memory, addr)
+                    .unwrap();
+                println!("OPA abort with message: {:?}", msg);
+                std::process::exit(1);
+            },
+        );
+        linker.define("env", "opa_abort", opa_abort)?;
```
Comment (flavio) [inline crates/burrego/src/opa/wasm.rs:127]: Unrelated with the PR, we will have to handle abort in different ways.

File: crates/burrego/src/opa/wasm.rs [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+        //Called to emit a message from the policy evaluation.
+        //The addr refers to a null-terminated string in the shared memory buffer.
+        let opa_println = Func::wrap(
+            &mut store,
+            move |mut caller: Caller<'_, Option<StackHelper>>, addr: i32| {
+                let stack_helper = caller.data().unwrap();
+                let msg = stack_helper
+                    .pull_json(caller.as_context_mut(), &memory, addr)
+                    .unwrap();
+                println!("Message coming from the policy: {:?}", msg);
```
Comment (flavio) [inline crates/burrego/src/opa/wasm.rs:139]: Same here, we will probably have to build a callback mechanism. The consumer will provide a function/closure to use to handle `abort` and `println`

File: src/policy_evaluator.rs [hunk 7]
```diff
@@ -110,26 +128,51 @@ impl PolicyEvaluator {
...
+                    |_| Ok(0),
+                    Policy::new,
+                    policy_execution_mode,
+                )?;
+                let evaluator = burrego::opa::wasm::Evaluator::new(id, &policy_contents)?;
+                let policy_runtime = Runtime::Burrego(BurregoEvaluator {
+                    evaluator,
+                    entrypoint_id: 0, // TODO: let kwctl/policy-server populate this
+                    input: serde_json::from_str("{}")?, // TODO: let kwctl/policy-server populate this
+                    data: serde_json::from_str("{}")?, // TODO: let kwctl/policy-server populate this
```
Comment (flavio) [inline src/policy_evaluator.rs:168]: nitpicking, I think it's nicer to use `json!({})`

**Review by flavio — CHANGES_REQUESTED** [review]
Thanks for the hard work! I've left some notes here and there.

I didn't elaborate on the `TODO`, but we will have to talk about them in a different way.

We also have to discuss about how to handle this work, I wouldn't merge that into `main` yet. I would rather do PRs against a dedicated branch and then, when everything is good, do a PR from that branch into `main`. What do you think?

File: crates/burrego/.github/workflows/tests.yml (hunk 2 above)
Comment (ereslibre) [reply crates/burrego/.github/workflows/tests.yml:1]: Let me check. If not, I can try some tricks like a symlink. I will report back.

File: crates/burrego/src/opa/evaluation_result.rs (hunk 4 above)
Comment (ereslibre) [reply crates/burrego/src/opa/evaluation_result.rs:6]: Yes, the main reason I didn't enter in detail here is: do we want to have this structure on the consumable crate, or our API is meant  to return `serde_json`? I prefer this approach, and I think I can expand this struct to the documented fields -- have to check where it's documented.

File: crates/burrego/.github/workflows/tests.yml (hunk 2 above)
Comment (ereslibre) [reply crates/burrego/.github/workflows/tests.yml:1]: A symlink does not work. We will have to have a targeted workflow for it in the root folder.

Comment (ereslibre) [reply crates/burrego/.github/workflows/tests.yml:1]: Opened https://github.com/kubewarden/policy-evaluator/issues/18

**Comment by ereslibre** [comment]
> We also have to discuss about how to handle this work, I wouldn't merge that into main yet. I would rather do PRs against a dedicated branch and then, when everything is good, do a PR from that branch into main. What do you think?

I think this is the best approach. We can discuss and keep improving it. We only have to be careful if it's more than one person touching the code and force pushing to a given shared branch.

**Comment by ereslibre** [comment]
Moving this PR base branch to the `policy-evaluator` repo inside the `kubewarden` org.

**Pull request closed without merging**